### ISSUE TITLE
treat directories as records

### ISSUE THREAD

**Issue by sellout** [dhall-lang/dhall-lang#139]
The idea being that you could do
```
let Prelude = https://.../Prelude/
```
instead of
```
let Prelude = https://.../Prelude/package.dhall
```
avoiding the need to duplicate directory structures as explicit records (except in the case where you want to provide more curated interfaces).

The big issue here is that it reinstates “mixed” records, previously removed by #120.

**Comment by Gabriel439** [comment]
I view this feature as *technically* orthogonal to whether or not mixed records are allowed but *in practice* it is more usable if mixed records are allowed.  I say *technically* orthogonal because presumably if you had a directory where one file was a record of terms and another file was a record of types it would just fail to type-check.  The idea is that this is implemented at the import resolution phase which is then followed by the type-checking phase.

I'm unlikely to go back to permitting mixed records of types and terms because I don't have a sound theoretical basis for that.  However, even so I think this is could be a good idea since there are a lot of use cases where you just want the user experience to be "add a file to this directory" and the directory is fill of nothing but terms.  It's also pretty straightforward to have a separate directory for types.

The main usability issue I can think of is that Dhall might accidentally fail on hidden files that litter directories which are neither Dhall expressions nor sub-directories.

I'm also unlikely to get to this until after the first complete draft of the standard import semantics.

**Comment by sellout** [comment]
One bit of “magic” that would be nice with this is that since they’re _not_ explicit records, they could be “pruned” in normalization, so that only the referenced parts of the directory are ever actually imported / depended on.

**Comment by sellout** [comment]
“it would just fail to type-check” – very good point.

**Comment by Gabriel439** [comment]
Well, I probably wouldn't be able to implement something like how Nix does it (i.e. be lazy and only load files that are actually used).  The reason why is that the current semantics require that import resolution has to strictly precede normalization and type-checking and the flow of information is one-way so you can't use the presence of absence of a field access to determine if a file/subdirectory needs to be loaded like in Nix.

**Comment by Gabriel439** [comment]
So in the context of #219 I decided that the correct solution is to do a minor variation on what you proposed here using a new `import` keyword instead of `let`.  In other words this:

```haskell
    import Prelude = https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0

in  Prelude.`List`.map
```

... would be equivalent to this:

```haskell
https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0/List/map
```

... where the imported directory acts like a virtual record where accessing fields translates to navigating the directory tree that you import.  Like you suggested, the record is "lazy" so the only files you retrieve are the record fields you access.  This virtual record only exists at import resolution time and the virtual record is not a valid value in isolation, meaning that this is not valid:

```haskell
    let Prelude = https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0

in  Prelude
```

... mainly to avoid accidentally failing on unexpected hidden files or accidentally importing secrets.

The interpreter also would correctly handle shadowing, meaning that this:

```haskell
    let Prelude = https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0

in  let Prelude = { List = { map = 1 } }

in  Prelude.`List`.map
```

... would be equivalent to:

```haskell
1
```

**Comment by Gabriel439** [comment]
Thinking about this a little more, I'll make one correction to the sketch of how this work: the virtual record would only verify that that the values you access are not directories.  In other words if you do:

```haskell
    import Foo = ./foo

in  Foo
```

That will succeed if `./foo` is a file.

Vice versa, if you do:

```haskell
    import Bar = ./bar

in  Bar.x
```

That will fail if `./bar/x` is a directory

**Comment by Profpatsch** [comment]
How do hashed imports come into play here? You cannot have a hash for a whole, “lazily loaded” directory tree.

**Comment by f-f** [comment]
@Profpatsch I believe we can have hashes on the single imports just fine, because the "directory" import is useless until you try to use a child import, on which you can put a hash. In other words, directory imports won't need to have a hash themselves. (if we want them to, the tradeoff is that they cannot be imported lazily)

**Comment by Gabriel439** [comment]
One thing that might reduce confusion is still a directory-like syntax for accessing the package's files.  In other words, something like this syntax:

```haskell
    import ./Prelude = https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0

in  ./Prelude/List/map
```

In other words the `import` keyword creates a virtual local directory of the same name

The nice advantage of that is that we don't need to worry about dealing with shadowing (which simplifies the standardization) and it's more obvious to the user how you would protect it with an integrity check because it looks just like a local import:

```haskell
    import ./Prelude = https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0

in  ./Prelude/List/map sha256:...
```

**Comment by Profpatsch** [comment]
This also means you lose the deduplication of the other approach (you still have to paste the full link into each file.
That might be what the user wants (it’s similar to how golang does its imports) or it might not. If you see these imports in a “one import maps to one name per project” way, putting them into a file might be the way to go.

Idea: can we combine these two approaches, so that the user can have per-file but still factor it out into one central import if they want?

**Comment by Gabriel439** [comment]
@Profpatsch: Let's do one step at a time for now.  I think having to specify the virtual directory once per file is tolerable, given that it's no more verbose than the equivalent idiom to import a package in a language like Haskell (i.e. `import Some.Package`), which you also have to do at least once per file.

**Comment by Profpatsch** [comment]
> I think having to specify the virtual directory once per file is tolerable, given that it's no more verbose than the equivalent idiom to import a package in a language like Haskell (i.e. import Some.Package)

The difference is that in Haskell, the module structure (`Some.Package`) is decoupled from the package version (which is handled by GHC package databases).

Concerning the combination of both variants, it seems like a logical extension of “including a file is the same as writing it in place” if I can do:

```
    import ./Prelude = ./vendor/prelude
in ./Prelude/List/map
```

where `./vendor/prelude` contains `https://raw.githubusercontent.com/dhall-lang/Prelude/v1.0.0`. :)

**Comment by Gabriel439** [comment]
@Profpatsch: I would still like to wait a while before following up with support for alias files.  I like to give new features time to bake to study how people use them in the wild before building on top of them.

**Comment by Gabriel439** [comment]
@sellout @Profpatsch:

So there are two problems that I've run into with both this proposal (treating directories as virtual records) and the other proposal (aliasing imports as virtual directories):

### Problem 1: Bloat in import-related `let` expressions

For example, suppose that you want to protect all your imports with semantic integrity checks.  You probably wouldn't want to have a semantic integrity check inline at each use site, so you would end up with a `let` definition for each imported value just to prevent the semantic integrity checks from polluting the main code.

Similarly, every import would have an obligatory prefix (either the virtual directory or the virtual record) and (usually) a suffix (the `.dhall` extension).  If you wanted to prevent these prefixes and suffixes from polluting the code you'd also need to do so using one `let` definition per import ahead of time.

This problem primarily affects package users.

### Problem 2: Not first class

There's another issue that proposals share which is that "packages" imported in this way are not first class because both packaging proposals occur during the import-resolution phase instead of the normalization.  For example, if you want to combine two packages into one you can't just use `∧` or if you want to restrict the symbols a package exports you can't use `.{x, y, z}`.

This primarily affects package authors because there are work-arounds (just create a new directory with a lot of small files wrapping files from the old package), but it's inconvenient.

### Current approach

Due to these issues, I'm more inclined to lean more heavily on improving the usability of records for packaging (i.e. the `package.dhall` convention).  Give me some time (a week?) to think through some ideas for how to make them more usable so that users don't need to import separate records for types and terms.  I think if we can make them work better it would greatly improve ergonomics because then users would only need at most one `let` definition for each package and then packages would be actually composable as ordinary expressions.

**Comment by Gabriel439** [comment]
So I think the most promising approach is this pull request to add a new `Sort` above `Kind` in the type hierarchy:

https://github.com/dhall-lang/dhall-haskell/pull/563

This will allow us to at least consolidate all type-level exports into a single record (i.e. type constructors of all kinds will be able to coexist within the same record, which is an improvement over the current situation), but still will not permit mixing types and terms in a single record.  In other words, a package will typically "export" two records (assuming no kind-level programming).

It seems like there isn't a good way to handle arbitrary mixed records of types and terms in a way that doesn't require exposing explicit universe polymorphism to the end user.  See these two comments for more details:

* https://github.com/dhall-lang/dhall-haskell/pull/563#issuecomment-419556829
* https://github.com/dhall-lang/dhall-haskell/pull/563#issuecomment-419755783

So I think we should just adopt a `terms.dhall` and `types.dhall` idiom for Dhall packages and document the idiom in the wiki.  I think users will tolerate requiring two exports and if the idiom becomes used widely enough we could even add language support down the road for importing both records using a single URL/path.

**Comment by Gabriel439** [comment]
So just updating this ticket to reiterate that the solution I'm going with is exporting two records, one for terms and types.  The only thing I'm waiting on before closing this is a resolution to #250

**Comment by ocharles** [comment]
> So just updating this ticket to reiterate that the solution I'm going with is exporting two records, one for terms and types.

We're about to get mixed records again - how does this change your proposal?

**Comment by Gabriel439** [comment]
@ocharles: Yeah, once we get mixed records we should consolidate things just down into a `package.dhall` like the Prelude already does

**Comment by ocharles** [comment]
Agreed, In that case I think this can for now be closed as a `wontfix`. I don't think we need anything else open as we're all on the same page about a "unified" `package.dhall` for things.
